### PR TITLE
enhance(aws-ec2): Add support for terminating node of a cluster with self-managed nodegroup

### PR DESCRIPTION
**Pull request by uditgaurav**
Signed-off-by: udit <udit@chaosnative.com>

## Details

- This PR adds support for terminating nodes of a cluster with managed nodegroup. In a cluster with managed nodegroup when the instance is stopped by the experiment, the nodegroup waits for very few seconds (around 30s) and brings a new node (ec2-instance) for the cluster and terminates the older instance which was in the stopped state.
- A new check in the pre and post chaos section has been introduced which will check when the `MANAGED_NODEGROUP` env in the chaosengine is set to` enable` . The Pre Chaos check will check the status of the node whose instance id is provided in the engine and in post chaos check the total number of Ready nodes are verified after the addition of a new node during chaos injection.

Note: 
- When  `MANAGED_NODEGROUP` is enabled then the experiment will not try to start the instance instead it will check of the addition of the new node instance.
- The application on the node under chaos can be lost as it will get terminated we need to drain in as a pre-requisite.

### CODE REVIEW

**Comment by uditgaurav** [comment]
Experiment Logs: 

```bash
INFO[2021-02-22T23:10:08+05:30] Experiment Name: ec2-terminate               
INFO[2021-02-22T23:10:08+05:30] [PreReq]: Getting the ENV for the  experiment 
INFO[2021-02-22T23:10:08+05:30] [PreReq]: Updating the chaos result of ec2-terminate experiment (SOT) 
INFO[2021-02-22T23:10:09+05:30] The application information is as follows     Namespace=default Label="run=nginx" Chaos Duration=80 Ramp Time=0
INFO[2021-02-22T23:10:10+05:30] [Info]: Total number active nodes are: 2     
INFO[2021-02-22T23:10:10+05:30] [Status]: Getting the status of target node  
INFO[2021-02-22T23:10:11+05:30] The Node status are as follows                Node=ip-172-20-48-193.us-west-2.compute.internal Ready=true
INFO[2021-02-22T23:10:11+05:30] [Status]: Verify that the AUT (Application Under Test) is running (pre-chaos) 
INFO[2021-02-22T23:10:11+05:30] [Status]: Checking whether application containers are in ready state 
INFO[2021-02-22T23:10:30+05:30] [Status]: The Container status are as follows  Pod=nginx Readiness=true container=nginx
INFO[2021-02-22T23:10:32+05:30] [Status]: Checking whether application pods are in running state 
INFO[2021-02-22T23:10:32+05:30] [Status]: The status of Pods are as follows   Pod=nginx Status=Running
INFO[2021-02-22T23:10:36+05:30] [Status]: EC2 instance is in running state   
INFO[2021-02-22T23:10:36+05:30] [Chaos]: Stoping the desired EC2 instance    
INFO[2021-02-22T23:10:37+05:30] Stopping an ec2 instance:                     CurrentState=stopping PreviousState=running InstanceId=i-0d1ec4e395ee29a7b
INFO[2021-02-22T23:10:37+05:30] [Wait]: Wait for EC2 instance to come in stopped state 
INFO[2021-02-22T23:10:37+05:30] [Status]: Checking EC2 instance status       
INFO[2021-02-22T23:10:37+05:30] The instance state is stopping               
INFO[2021-02-22T23:10:40+05:30] The instance state is stopping               
INFO[2021-02-22T23:10:42+05:30] The instance state is stopping               
INFO[2021-02-22T23:10:45+05:30] The instance state is stopping               
INFO[2021-02-22T23:10:47+05:30] The instance state is stopping               
INFO[2021-02-22T23:10:50+05:30] The instance state is stopping               
INFO[2021-02-22T23:10:52+05:30] The instance state is stopping               
INFO[2021-02-22T23:10:54+05:30] The instance state is stopping               
INFO[2021-02-22T23:10:57+05:30] The instance state is stopping               
INFO[2021-02-22T23:10:59+05:30] The instance state is stopping               
INFO[2021-02-22T23:11:02+05:30] The instance state is stopping               
INFO[2021-02-22T23:11:04+05:30] The instance state is stopping               
INFO[2021-02-22T23:11:06+05:30] The instance state is stopping               
INFO[2021-02-22T23:11:09+05:30] The instance state is stopping               
INFO[2021-02-22T23:11:11+05:30] The instance state is stopping               
INFO[2021-02-22T23:11:14+05:30] The instance state is stopping               
INFO[2021-02-22T23:11:16+05:30] The instance state is stopping               
INFO[2021-02-22T23:11:19+05:30] The instance state is stopping               
INFO[2021-02-22T23:11:21+05:30] The instance state is stopping               
INFO[2021-02-22T23:11:23+05:30] The instance state is stopping               
INFO[2021-02-22T23:11:26+05:30] The instance state is stopping               
INFO[2021-02-22T23:11:28+05:30] The instance state is stopping               
INFO[2021-02-22T23:11:31+05:30] The instance state is stopping               
INFO[2021-02-22T23:11:33+05:30] The instance state is stopping               
INFO[2021-02-22T23:11:35+05:30] The instance state is stopping               
INFO[2021-02-22T23:11:38+05:30] The instance state is stopping               
INFO[2021-02-22T23:11:40+05:30] The instance state is stopped                
INFO[2021-02-22T23:11:42+05:30] [Wait]: Waiting for chaos duration of 80s before starting the instance 
INFO[2021-02-22T23:13:02+05:30] [Confirmation]: EC2 terminate chaos has been injected successfully 
INFO[2021-02-22T23:13:03+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:13:06+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:13:08+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:13:11+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:13:13+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:13:15+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:13:18+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:13:20+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:13:22+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:13:24+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:13:27+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:13:29+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:13:32+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:13:34+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:13:36+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:13:39+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:13:41+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:13:43+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:13:45+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:13:48+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:13:50+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:13:53+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:13:55+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:13:57+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:14:00+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:14:02+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:14:04+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:14:07+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:14:09+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:14:11+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:14:14+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:14:16+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:14:18+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:14:21+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:14:23+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:14:25+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:14:28+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:14:30+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:14:33+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:14:35+05:30] [Info]: Total number active nodes are: 1     
INFO[2021-02-22T23:14:37+05:30] [Info]: Total number active nodes are: 2     
INFO[2021-02-22T23:14:37+05:30] [Info]: The active node count is: 2          
INFO[2021-02-22T23:14:39+05:30] [Status]: Verify that the AUT (Application Under Test) is running (post-chaos) 
INFO[2021-02-22T23:14:39+05:30] [Status]: Checking whether application containers are in ready state 
INFO[2021-02-22T23:15:38+05:30] [Status]: The Container status are as follows  Pod=nginx Readiness=true container=nginx
INFO[2021-02-22T23:15:40+05:30] [Status]: Checking whether application pods are in running state 
INFO[2021-02-22T23:15:41+05:30] [Status]: The status of Pods are as follows   Pod=nginx Status=Running
INFO[2021-02-22T23:15:43+05:30] [The End]: Updating the chaos result of ec2-terminate experiment (EOT) 
```

File: pkg/cloud/aws/ec2-instance-status.go [hunk 1]
```diff
@@ -83,3 +89,65 @@ func GetRandomInstance(region string, session *session.Session) (string, error)
...
+			if experimentsDetails.ActiveNodes != activeNodes {
+				return errors.Errorf("fail to get equal active node post chaos")
+			}
+			return nil
+		})
+	return err
+}
+
+// getTargetNodeAndTotalNodeCount fetch the target node and total node count from the cluster
+func getTargetNodeAndTotalNodeCount(ec2InstanceID string, clients clients.ClientSets) (int, error) {
```
Comment (ispeakc0de) [inline pkg/cloud/aws/ec2-instance-status.go:133]: Can we have a meaningful name? i.e `getActiveNodeCount`

File: pkg/cloud/aws/ec2-instance-status.go [hunk 2]
```diff
@@ -83,3 +89,65 @@ func GetRandomInstance(region string, session *session.Session) (string, error)
...
+// PreChaosNodeStatusCheck fetch the target node name from instance id and checks its status also fetch the total active nodes in the cluster
+func PreChaosNodeStatusCheck(experimentsDetails *experimentTypes.ExperimentDetails, clients clients.ClientSets) error {
+
+	nodeList, err := clients.KubeClient.CoreV1().Nodes().List(metav1.ListOptions{})
+	if err != nil {
+		return errors.Errorf("fail to get the nodes, err: %v", err)
+	}
+	for _, node := range nodeList.Items {
+		err = status.CheckNodeStatus(node.Name, experimentsDetails.Timeout, experimentsDetails.Delay, clients)
+		log.Infof("[Info]: The cluster is unhealthy this might not work, due to %v", err)
```
Comment (ispeakc0de) [inline pkg/cloud/aws/ec2-instance-status.go:102]: Can we print this error only if err is not nil? and return the err as well

Comment (uditgaurav) [reply pkg/cloud/aws/ec2-instance-status.go]: Have added the condition. We don't want to return it. We just wanted to inform the user that your cluster is not healthy pre-pre chaos. That's it.

File: pkg/cloud/aws/ec2-instance-status.go (hunk 1 above)
Comment (uditgaurav) [reply pkg/cloud/aws/ec2-instance-status.go]: done!

File: chaoslib/litmus/ec2-terminate/lib/ec2-terminate.go [hunk 3]
```diff
@@ -164,7 +166,7 @@ func WaitForEC2Down(experimentsDetails *experimentTypes.ExperimentDetails) error
 			if err != nil {
 				return errors.Errorf("failed to get the instance status")
 			}
-			if instanceState != "stopped" {
+			if (experimentsDetails.ManagedNodegroup != "enable" && instanceState != "stopped") || (experimentsDetails.ManagedNodegroup == "enable" && instanceState == "terminated") {
```
Comment (ispeakc0de) [inline chaoslib/litmus/ec2-terminate/lib/ec2-terminate.go:169]:
```suggestion
			if (experimentsDetails.ManagedNodegroup != "enable" && instanceState != "stopped") || (experimentsDetails.ManagedNodegroup == "enable" && instanceState != "terminated") {
```

**Review by ispeakc0de — DISMISSED** [review]

**Comment by uditgaurav** [comment]
Updated Logs:

```
W0315 10:22:11.713856       1 client_config.go:541] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
time="2021-03-15T10:22:11Z" level=info msg="Experiment Name: ec2-terminate"
time="2021-03-15T10:22:11Z" level=info msg="[PreReq]: Getting the ENV for the  experiment"
time="2021-03-15T10:22:11Z" level=info msg="[PreReq]: Updating the chaos result of ec2-terminate experiment (SOT)"
time="2021-03-15T10:22:11Z" level=info msg="The application information is as follows" Namespace=default Label= Chaos Duration=60 Ramp Time=0
time="2021-03-15T10:22:11Z" level=info msg="The Node status are as follows" Ready=true Node=ip-192-168-16-41.us-west-2.compute.internal
time="2021-03-15T10:22:11Z" level=info msg="The Node status are as follows" Ready=true Node=ip-192-168-80-202.us-west-2.compute.internal
time="2021-03-15T10:22:11Z" level=info msg="[Info]: Total number active nodes are: 2"
time="2021-03-15T10:22:11Z" level=info msg="[Status]: Verify that the AUT (Application Under Test) is running (pre-chaos)"
time="2021-03-15T10:22:11Z" level=info msg="[Status]: Checking whether application containers are in ready state"
time="2021-03-15T10:22:13Z" level=info msg="[Status]: The Container status are as follows" container=ec2-terminate-dplpq0 Pod=ec2-terminate-dplpq0-j267f Readiness=true
time="2021-03-15T10:22:13Z" level=info msg="[Status]: The Container status are as follows" Pod=nginx-86c57db685-85ktc Readiness=true container=nginx
time="2021-03-15T10:22:13Z" level=info msg="[Status]: The Container status are as follows" container=chaos-runner Pod=nginx-chaos-runner Readiness=true
time="2021-03-15T10:22:15Z" level=info msg="[Status]: Checking whether application pods are in running state"
time="2021-03-15T10:22:15Z" level=info msg="[Status]: The status of Pods are as follows" Pod=ec2-terminate-dplpq0-j267f Status=Running
time="2021-03-15T10:22:15Z" level=info msg="[Status]: The status of Pods are as follows" Pod=nginx-86c57db685-85ktc Status=Running
time="2021-03-15T10:22:15Z" level=info msg="[Status]: The status of Pods are as follows" Pod=nginx-chaos-runner Status=Running
time="2021-03-15T10:22:18Z" level=info msg="[Status]: EC2 instance is in running state"
time="2021-03-15T10:22:18Z" level=info msg="[Chaos]: Stoping the desired EC2 instance"
time="2021-03-15T10:22:18Z" level=info msg="Stopping an ec2 instance:" CurrentState=stopping PreviousState=running InstanceId=i-0dd5b566ff6a6e3e7
time="2021-03-15T10:22:18Z" level=info msg="[Wait]: Wait for EC2 instance to come in stopped state"
time="2021-03-15T10:22:18Z" level=info msg="[Status]: Checking EC2 instance status"
time="2021-03-15T10:22:18Z" level=info msg="The instance state is running"
time="2021-03-15T10:22:20Z" level=info msg="The instance state is stopping"
time="2021-03-15T10:22:22Z" level=info msg="The instance state is stopping"
time="2021-03-15T10:22:24Z" level=info msg="The instance state is stopping"
time="2021-03-15T10:22:27Z" level=info msg="The instance state is stopping"
time="2021-03-15T10:22:29Z" level=info msg="The instance state is stopping"
time="2021-03-15T10:22:31Z" level=info msg="The instance state is stopping"
time="2021-03-15T10:22:33Z" level=info msg="The instance state is stopping"
time="2021-03-15T10:22:35Z" level=info msg="The instance state is stopping"
time="2021-03-15T10:22:37Z" level=info msg="The instance state is stopping"
time="2021-03-15T10:22:40Z" level=info msg="The instance state is stopping"
time="2021-03-15T10:22:42Z" level=info msg="The instance state is stopping"
time="2021-03-15T10:22:44Z" level=info msg="The instance state is stopping"
time="2021-03-15T10:22:46Z" level=info msg="The instance state is stopping"
time="2021-03-15T10:22:48Z" level=info msg="The instance state is stopping"
time="2021-03-15T10:22:50Z" level=info msg="The instance state is stopping"
time="2021-03-15T10:22:52Z" level=info msg="The instance state is stopping"
time="2021-03-15T10:22:55Z" level=info msg="The instance state is stopping"
time="2021-03-15T10:22:57Z" level=info msg="The instance state is stopping"
time="2021-03-15T10:22:59Z" level=info msg="The instance state is stopping"
time="2021-03-15T10:23:01Z" level=info msg="The instance state is stopping"
time="2021-03-15T10:23:03Z" level=info msg="The instance state is stopping"
time="2021-03-15T10:23:05Z" level=info msg="The instance state is stopping"
time="2021-03-15T10:23:08Z" level=info msg="The instance state is stopping"
time="2021-03-15T10:23:10Z" level=info msg="The instance state is stopping"
time="2021-03-15T10:23:12Z" level=info msg="The instance state is stopping"
time="2021-03-15T10:23:14Z" level=info msg="The instance state is stopping"
time="2021-03-15T10:23:16Z" level=info msg="The instance state is stopping"
time="2021-03-15T10:23:18Z" level=info msg="The instance state is stopping"
time="2021-03-15T10:23:20Z" level=info msg="The instance state is stopping"
time="2021-03-15T10:23:22Z" level=info msg="The instance state is stopping"
time="2021-03-15T10:23:25Z" level=info msg="The instance state is stopping"
time="2021-03-15T10:23:27Z" level=info msg="The instance state is stopped"
time="2021-03-15T10:23:29Z" level=info msg="The instance state is stopped"
time="2021-03-15T10:23:31Z" level=info msg="The instance state is stopped"
time="2021-03-15T10:23:33Z" level=info msg="The instance state is stopped"
time="2021-03-15T10:23:35Z" level=info msg="The instance state is stopped"
time="2021-03-15T10:23:38Z" level=info msg="The instance state is stopped"
time="2021-03-15T10:23:40Z" level=info msg="The instance state is stopped"
time="2021-03-15T10:23:42Z" level=info msg="The instance state is stopped"
time="2021-03-15T10:23:44Z" level=info msg="The instance state is shutting-down"
time="2021-03-15T10:23:46Z" level=info msg="The instance state is shutting-down"
time="2021-03-15T10:23:48Z" level=info msg="The instance state is terminated"
time="2021-03-15T10:23:50Z" level=info msg="[Wait]: Waiting for chaos duration of 60s before starting the instance"
time="2021-03-15T10:24:50Z" level=info msg="[Confirmation]: EC2 terminate chaos has been injected successfully"
time="2021-03-15T10:24:50Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:24:52Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:24:54Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:24:56Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:24:58Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:25:00Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:25:02Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:25:04Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:25:06Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:25:08Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:25:10Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:25:12Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:25:14Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:25:16Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:25:18Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:25:20Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:25:22Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:25:24Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:25:27Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:25:29Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:25:31Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:25:33Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:25:35Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:25:37Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:25:39Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:25:41Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:25:43Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:25:45Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:25:47Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:25:49Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:25:51Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:25:53Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:25:55Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:25:57Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:25:59Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:26:01Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:26:03Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:26:05Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:26:07Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:26:09Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:26:11Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:26:13Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:26:15Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:26:17Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:26:19Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:26:21Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:26:23Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:26:25Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:26:27Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:26:29Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:26:31Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:26:33Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:26:35Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:26:37Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:26:39Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:26:41Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:26:43Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:26:45Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:26:47Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:26:49Z" level=info msg="[Info]: Total number active nodes are: 1"
time="2021-03-15T10:26:51Z" level=info msg="[Info]: Total number active nodes are: 2"
time="2021-03-15T10:26:53Z" level=info msg="[Status]: Verify that the AUT (Application Under Test) is running (post-chaos)"
time="2021-03-15T10:26:53Z" level=info msg="[Status]: Checking whether application containers are in ready state"
time="2021-03-15T10:26:53Z" level=info msg="[Status]: The Container status are as follows" Pod=ec2-terminate-dplpq0-j267f Readiness=true container=ec2-terminate-dplpq0
time="2021-03-15T10:26:53Z" level=info msg="[Status]: The Container status are as follows" Pod=nginx-86c57db685-85ktc Readiness=true container=nginx
time="2021-03-15T10:26:53Z" level=info msg="[Status]: The Container status are as follows" Pod=nginx-chaos-runner Readiness=true container=chaos-runner
time="2021-03-15T10:26:55Z" level=info msg="[Status]: Checking whether application pods are in running state"
time="2021-03-15T10:26:55Z" level=info msg="[Status]: The status of Pods are as follows" Pod=ec2-terminate-dplpq0-j267f Status=Running
time="2021-03-15T10:26:55Z" level=info msg="[Status]: The status of Pods are as follows" Pod=nginx-86c57db685-85ktc Status=Running
time="2021-03-15T10:26:55Z" level=info msg="[Status]: The status of Pods are as follows" Pod=nginx-chaos-runner Status=Running
time="2021-03-15T10:26:57Z" level=info msg="[The End]: Updating the chaos result of ec2-terminate experiment (EOT)"
```

**Comment by uditgaurav** [comment]
We need to cordon the target node pre chaos and shift (drain) if any application is running on the target node otherwise the application can get lost.

**Review by ispeakc0de — APPROVED** [review]

**Review by ksatchit — APPROVED** [review]

**Pull request merged**